core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -341,4 +385,63 @@ object ZLayer {
    * Constructs a layer from the specified value.
    */
   def succeed[A: Tagged](a: => A): ZLayer.NoDeps[Nothing, Has[A]] = ZLayer(ZManaged.succeed(Has(a)))
+
+  /**
+   * A `MemoMap` memoizes dependencies.
+   */
+  private trait MemoMap { self =>
+
+    /**
+     * Retrieves a dependency from the memo map if it exists.
+     */
+    def get[E, A, B <: Has[_]](layer: ZLayer[A, E, B]): UIO[Option[B]]
+
+    /**
+     * Stores a dependency in the memo map.
+     */
+    def memoize[E, A, B <: Has[_]](layer: ZLayer[A, E, B], b: B): UIO[Unit]
+
+    /**
+     * Checks the memo map to see if a dependency exists. If it is, immediately
+     * returns it. Otherwise, obtains the dependency, stores it in the memo map,
+     * and adds a finalizer to the outer `Managed`.
+     */
+    final def getOrElseMemoize[E, A, B <: Has[_]](
+      layer: ZLayer[A, E, B],
+      finalizerRef: Ref[Exit[Any, Any] => ZIO[Any, Nothing, Any]]
+    ): ZManaged[A, E, B] =
+      ZManaged {
+        get(layer).flatMap {
```

Comment:
Is this safe for arbitrary interruption, or would it be useful to add `uninterruptibleMask` to protect critical regions, and allow interruption only some places (e.g. `reservation.acquire`?)?

Reply:
So I think we are okay here but it is definitely the right question. My analysis is:

1. First we check if the dependency is in the map and if so just return it. I don't think there is any problem with being interrupted here because we haven't acquired any resources at this point that require finalization.
2. If we don't have the resource yet, we do `layer.scope.flatMap(_.apply(self))` which creates one Managed that describes entering both the inner and outer scopes to create the dependency.
3. Then we call `reserve` on it to access the `acquire` and `release` actions of the managed separately but we still haven't actually acquired any resources.
4. The actual acquisition of resources comes in `reservation.acquire` which is in this reservation we are constructing that will create a new `Managed`, so the acquisition of the dependency and adding it to the memo map will be uninterruptible and if it starts the finalizers are guaranteed to be added to the outer scope, also uninterruptibly, because that is in the release action of the managed.

Reply:
It's tricky, but I think you're right..